### PR TITLE
CODAP-1469: make residual-plot points marquee-selectable

### PR DESCRIPTION
**Pull request by kswenson**
## Summary

Adds marquee (drag-rectangle) selection to the residual plot, alongside the
existing click selection. Fixes CODAP-1469.

- Dragging on the residual strip selects the residual points inside the rectangle;
  selection updates dynamically while dragging, and Shift adds to the current
  selection.
- Selection is confined to the residual strip (separate zone) — the upper-plot
  marquee and the map marquee are unchanged. A rectangle spanning both the main
  Y-axis space and the residual space has no coherent meaning, so the two zones
  stay independent.
- A click (no drag) still deselects all cases, as before.

## Implementation notes

- Residual points are hand-drawn SVG circles that the point renderer doesn't
  track, so the shared `Background` marquee (which hit-tests the renderer's
  R-tree) can't see them. The new `useResidualMarquee` hook builds its own
  R-tree from the residual positions on pointerdown and selects via the same
  delta hit-testing — the R-tree is built up front to stay efficient with large
  datasets (one residual point per case).
- Refactor: the residual rendering/sync code was extracted out of
  `scatter-plot.tsx` into a `useResidualPlot` hook, and `Background`'s
  `getCasesForDelta` was moved into `data-display-utils.ts` so both marquees
  share it.

## Reviewer notes

This PR is a mix of new logic and mechanical moves — the commit history is
structured so each is reviewable on its own.

**New code (warrants closer review):**
- `use-residual-marquee.ts` — the new drag-select behavior: R-tree build,
  delta hit-testing, dynamic/Shift selection, the pointer→plot-area coordinate
  conversion, and the click-vs-drag threshold that preserves click-to-deselect.
- `residual-marquee-utils.ts` — new pure helpers (`buildResidualPositions`,
  `clampRectToLowerRegion`) + their unit tests.
- `scatter-plot.tsx` — the *new* bits only: the `useResidualMarquee` call and
  the residual hit-rect's `onPointerDown`/`onClick` wiring (the rest of the diff
  in this file is the extraction below).
- `use-marquee-state.ts` + the `MarqueeStateContext` provider in `graph.tsx` —
  new, but trivial context plumbing.
- `bivariate-adornments.spec.ts` — new Cypress test.
- the `getCasesForDelta` unit test in `data-display-utils.test.ts`.

**Moved / mechanical (lighter review):**
- `use-residual-plot.ts` — a behavior-preserving extraction of the residual
  rendering/sync code out of `scatter-plot.tsx` (no logic changes; the guards
  and `untracked` wrapping were moved verbatim). Most of the `scatter-plot.tsx`
  deletions are the other side of this move.
- `getCasesForDelta` moved from `background.tsx` into `data-display-utils.ts`;
  `background.tsx` just imports it now.
- the `graph.tsx` reindent commit is whitespace only (JSX nested under the new
  provider).

## Testing

- Unit tests for the extracted `getCasesForDelta` and the pure residual-marquee
  geometry helpers.
- Cypress e2e: drag-selecting residual points on the strip (plus the existing
  click select/deselect coverage).

### CODE REVIEW

File: v3/src/components/graph/plots/scatter-plot/use-residual-marquee.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      if (draggingRef.current) {
+        suppressClickRef.current = true // the trailing click must not deselect what the drag selected
+        const numCases = dataset.selection.size
+        if (numCases > 0) Logger.log(`marqueeSelection: ${numCases}`, { numCases }, "plot")
+      }
+    }
+
+    window.addEventListener("pointermove", onMove)
+    window.addEventListener("pointerup", onUp)
+  }, [adornmentsStore, dataConfiguration, dataset, layout, marqueeState])
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/plots/scatter-plot/use-residual-marquee.ts:116]: Good catch! Added a `pointercancel` listener that ends the drag the same way `pointerup` does — removing the window listeners and clearing the marquee rect — so a touch/gesture cancellation no longer leaves them attached.

File: v3/src/components/data-display/data-display-utils.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import RTreeLib from "rtree"
 import {measureText} from "../../hooks/use-measure-text"
 import {IDataSet} from "../../models/data/data-set"
 import { selectCases, setOrExtendSelection } from "../../models/data/data-set-utils"
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/data-display-utils.ts]: Good point, and timely given the bundle-optimization work. `getCasesForDelta` only needs the tree's type, never constructs one — so I replaced `ReturnType<typeof RTreeLib>` with a minimal `RTree` interface (`search(rect): CaseObject[]`) and dropped the runtime `rtree` import from `data-display-utils.ts`. Tree construction stays in the callers (`background.tsx` and the residual-marquee hook) that already depend on `rtree`.

File: v3/src/components/graph/plots/scatter-plot/use-residual-marquee.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
+import { ScaleLinear } from "d3"
+import { useCallback, useRef } from "react"
+import RTreeLib from "rtree"
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/plots/scatter-plot/use-residual-marquee.ts:3]: Thanks for flagging this, but it compiles as-is — `build:tsc` passes. `@types/react` declares a global `React` namespace (`export as namespace React`), so `React.PointerEvent`/`React.MouseEvent` resolve without importing `React`. Several existing files use this same pattern (e.g. `slider-component.tsx`, `use-tile-drag.ts`, `web-view.tsx`), so I've kept it consistent with the codebase.

File: v3/src/components/graph/plots/scatter-plot/use-residual-marquee.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    const rectBounds = event.currentTarget.getBoundingClientRect()
+    startRef.current = {
+      x: event.clientX - rectBounds.left,
+      y: event.clientY - rectBounds.top + layout.plotHeight
+    }
+    sizeRef.current = { w: 0, h: 0 }
+    prevRectRef.current = undefined
+    needsClearRef.current = !event.shiftKey
+    draggingRef.current = false
+
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/plots/scatter-plot/use-residual-marquee.ts]: Good catch — real bug. If a drag ended without a trailing click on the rect (pointerup off the rect), `suppressClickRef` stayed true and would swallow the next genuine click. Now reset at the start of each `onPointerDown`, so suppression only applies to the click immediately following a drag.

**Comment by bfinzer** [comment]
## Code Review: PR #2677 — CODAP-1469: Residual-plot marquee selection

### Overview

Adds drag-marquee selection to the residual strip alongside existing click behavior. Selection is confined to the strip (independent of the upper-plot marquee, per the "two zones with no coherent combined meaning" rationale). Also does substantial refactoring: extracts residual rendering out of `scatter-plot.tsx` into `use-residual-plot.ts` (~180 lines moved), and lifts `getCasesForDelta` from `background.tsx` into shared `data-display-utils.ts`. Adds a `MarqueeStateContext` for cross-component sharing of the display marquee state.

### Strengths

- **Excellent commit hygiene and PR description.** The reviewer notes explicitly separate "new logic" from "mechanical moves," and each commit is reviewable independently.
- **Clear documentation of load-bearing patterns.** The `syncResidualPlot` autorun preserves the very useful "read-and-compare guard" warning verbatim (`use-residual-plot.ts:1024-1030`). Similarly the plot-area coordinate anchoring is well-explained (`use-residual-marquee.ts:55-58`).
- **Delta hit-testing shared correctly.** `getCasesForDelta` extracted cleanly to `data-display-utils.ts:259` with a matching unit test.
- **Pure helpers with tests.** `buildResidualPositions` / `clampRectToLowerRegion` are well-isolated and covered.
- **Good coverage of edge cases** in `useResidualMarquee`: click-vs-drag threshold, `pointercancel` cleanup, stale suppression clear on new pointerdown, shift-additive selection, empty-selection log skip.

### Suggestions / Minor Issues

**1. `RTree` interface is a partial type lie.** (`data-display-utils.ts:252`)
```ts
export interface RTree {
  search(rect: rTreeRect): CaseObject[]
}
```
The actual `rtree` library's `search` returns whatever was inserted (`any[]`). The narrow type is practically fine because both callers insert `CaseObject`s, but it hides that the tree could hold anything else. Consider making it generic (`RTree<T = CaseObject>`) or adding a comment noting the caller-side invariant.

**2. `sizeRef` accumulation is more complex than needed.** (`use-residual-marquee.ts:77-80`)
```ts
sizeRef.current.w += moveEvent.clientX - prev.x
sizeRef.current.h += moveEvent.clientY - prev.y
prev.x = moveEvent.clientX
prev.y = moveEvent.clientY
```
This telescopes to `moveEvent.clientX - event.clientX` (net displacement from pointerdown). The `prev` tracking and accumulation form is equivalent to `sizeRef.current.w = moveEvent.clientX - startClientX`. Not a bug, just an opportunity to simplify.

**3. `ILowerRegion` assumes `left = 0` implicitly.** (`residual-marquee-utils.ts:26-30`, `use-residual-marquee.ts:73`)
The interface has `{ top, bottom, right }` with no `left`. That's fine given the residual hit-rect is drawn at `x={0}`, but adding `left` (or a comment noting the assumption) would make the constraint explicit and would head off future breakage if the plot origin ever moves.

**4. Test coverage gaps.** (`bivariate-adornments.spec.ts:425`)
The e2e covers a basic drag-select but doesn't exercise:
- **Shift-drag additive behavior** — explicitly called out in the PR summary as new behavior.
- **Post-drag click-to-deselect** — the "trailing click must not deselect what the drag selected" suppression is subtle and worth an explicit assertion.
- **The marquee visual rect** actually being drawn during drag (the test only checks fill colors on points).

**5. Multi-touch race.** If a second `pointerdown` fires on the residual rect before an in-flight drag ends, it will overwrite `startRef`/`treeRef` and install a second set of window listeners, all sharing the mutated refs. Background likely has the same limitation, so this is not a blocker, but worth noting for touch scenarios.

**6. Import ordering.** (`data-display-utils.test.ts:109`)
```ts
import RTreeLib from "rtree"
```
Wedged between local imports rather than grouped with other package imports at the top. Minor.

### Correctness Concerns — None Blocking

- Coordinate transform (`use-residual-marquee.ts:59-63`) correctly anchors to the hit rect's own `getBoundingClientRect()`, so it's robust to axis-margin changes.
- The `treeRef.current` narrowing (`ReturnType<typeof RTreeLib>` → `RTree`) at the `getCasesForDelta` call site is structurally compatible.
- The `useResidualPlot` extraction is behavior-preserving (guards, `untracked` wrapping, and the post-mount effect all move verbatim).
- Click suppression state is correctly reset at each new `pointerdown` so a drag ending off-target doesn't poison the next click.

### Verdict

Ship it. The refactor is clean, the new logic is well-thought-out with good defensive handling, and the diff is unusually well-organized for review. The suggestions above are polish, not blockers — I'd particularly encourage adding a Shift-drag e2e assertion before merge since that's an explicit new behavior.

**Review by bfinzer — APPROVED** [review]
👍🏻LGTM
See Claude's review in comment
